From cf5148113c66c8664948596ca68c7141879e772a Mon Sep 17 00:00:00 2001
From: sunshineinabox <aqemail@gmail.com>
Date: Tue, 22 Sep 2026 11:26:57 -0700
Subject: [PATCH] wsi/wayland: support rotated preTransform

Advertise the 90, 180 and 270 degree transforms and pass preTransform on
with wl_surface.set_buffer_transform, so an app that renders rotated
content can have it scanned out without the compositor rotating it.

currentTransform stays identity. The app owns the wl_surface, so only the
app gets wl_surface.preferred_buffer_transform.

Damage rects are in the untransformed space, so rotated presents damage
the whole buffer.
---
 src/vulkan/wsi/wsi_common_wayland.c | 37 +++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/src/vulkan/wsi/wsi_common_wayland.c b/src/vulkan/wsi/wsi_common_wayland.c
index 20050922a1b..a6f67b8ed9a 100644
--- a/src/vulkan/wsi/wsi_common_wayland.c
+++ b/src/vulkan/wsi/wsi_common_wayland.c
@@ -187,6 +187,9 @@ struct wsi_wl_surface {
    struct zwp_linux_dmabuf_feedback_v1 *wl_dmabuf_feedback;
    struct dmabuf_feedback dmabuf_feedback, pending_dmabuf_feedback;
    struct vk_instance *instance;
+
+   /* last wl_surface buffer transform sent */
+   enum wl_output_transform buffer_transform;
 };
 
 struct wsi_wl_swapchain {
@@ -203,6 +206,8 @@ struct wsi_wl_swapchain {
    VkExtent2D extent;
    VkFormat vk_format;
    enum wsi_wl_buffer_type buffer_type;
+
+   enum wl_output_transform buffer_transform;
    uint32_t drm_format;
    enum wl_shm_format shm_format;
 
@@ -1853,6 +1858,22 @@ wsi_wl_surface_get_min_image_count(struct wsi_wl_display *display,
    return display->fifo_manager ? WSI_WL_DEFAULT_NUM_IMAGES : WSI_WL_BUMPED_NUM_IMAGES;
 }
 
+/* Vulkan rotations are clockwise, wl_output ones counter-clockwise */
+static enum wl_output_transform
+wsi_wl_transform_from_vk(VkSurfaceTransformFlagBitsKHR transform)
+{
+   switch (transform) {
+   case VK_SURFACE_TRANSFORM_ROTATE_90_BIT_KHR:
+      return WL_OUTPUT_TRANSFORM_270;
+   case VK_SURFACE_TRANSFORM_ROTATE_180_BIT_KHR:
+      return WL_OUTPUT_TRANSFORM_180;
+   case VK_SURFACE_TRANSFORM_ROTATE_270_BIT_KHR:
+      return WL_OUTPUT_TRANSFORM_90;
+   default:
+      return WL_OUTPUT_TRANSFORM_NORMAL;
+   }
+}
+
 static VkResult
 wsi_wl_surface_get_capabilities(VkIcdSurfaceBase *icd_surface,
                                 struct wsi_device *wsi_device,
@@ -1889,6 +1910,13 @@ wsi_wl_surface_get_capabilities(VkIcdSurfaceBase *icd_surface,
    };
 
    caps->surfaceCapabilities.supportedTransforms = VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR;
+   if (wl_proxy_get_version((struct wl_proxy *)surface->surface) >=
+       WL_SURFACE_SET_BUFFER_TRANSFORM_SINCE_VERSION) {
+      caps->surfaceCapabilities.supportedTransforms |=
+         VK_SURFACE_TRANSFORM_ROTATE_90_BIT_KHR |
+         VK_SURFACE_TRANSFORM_ROTATE_180_BIT_KHR |
+         VK_SURFACE_TRANSFORM_ROTATE_270_BIT_KHR;
+   }
    caps->surfaceCapabilities.currentTransform = VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR;
    caps->surfaceCapabilities.maxImageArrayLayers = 1;
 
@@ -3399,6 +3427,14 @@ wsi_wl_swapchain_queue_present(struct wsi_swapchain *wsi_chain,
    }
 
    assert(image_index < chain->base.image_count);
+   if (wsi_wl_surface->buffer_transform != chain->buffer_transform) {
+      wl_surface_set_buffer_transform(wsi_wl_surface->wayland_surface.wrapper,
+                                      chain->buffer_transform);
+      wsi_wl_surface->buffer_transform = chain->buffer_transform;
+   }
+   /* damage rects are not in the transformed buffer's coordinates */
+   if (chain->buffer_transform != WL_OUTPUT_TRANSFORM_NORMAL)
+      damage = NULL;
    wl_surface_attach(wsi_wl_surface->wayland_surface.wrapper,
                      chain->images[image_index].wayland_buffer.buffer, 0, 0);
 
@@ -3909,6 +3945,7 @@ wsi_wl_surface_create_swapchain(VkIcdSurfaceBase *icd_surface,
    VkPresentModeKHR present_mode = wsi_swapchain_get_present_mode(wsi_device, pCreateInfo);
 
    chain->color.colorspace = pCreateInfo->imageColorSpace;
+   chain->buffer_transform = wsi_wl_transform_from_vk(pCreateInfo->preTransform);
 
    enum wsi_wl_buffer_type buffer_type;
    struct wsi_base_image_params *image_params = NULL;
